MonoGame.Framework/Graphics/GraphicsDevice.cs

```diff
@@ -116,11 +116,16 @@ private bool PixelShaderDirty
 
         private readonly ConstantBufferCollection _vertexConstantBuffers = new ConstantBufferCollection(ShaderStage.Vertex, 16);
         private readonly ConstantBufferCollection _pixelConstantBuffers = new ConstantBufferCollection(ShaderStage.Pixel, 16);
+
+        /// <summary>
+        /// Effects can be added to the EffectCache from multiple threads.
+        /// </summary>
+        private readonly object _effectCacheLock = new object();
 
         /// <summary>
         /// The cache of effects from unique byte streams.
         /// </summary>
-        internal Dictionary<int, Effect> EffectCache;
+        private Dictionary<int, Effect> EffectCache;

```

Comment:
Why not use `ConcurrentDictionary` with [`GetOrAdd`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=netframework-4.8#System_Collections_Concurrent_ConcurrentDictionary_2_GetOrAdd__1__0_System_Func__0___0__1____0_)? Let the runtime handle things :)

Reply:
I don't mind the lock approach... simple, easy to debug, not dependent on .NET lib features.

Moving it to GraphicsDevice *i think* mirrors some of the other cleanups we have done in other caches right?

Reply:
Yeah I copied the same locking mechanism to match the coding standard used elsewhere in the codebase:

https://github.com/MonoGame/MonoGame/blob/bdc5951e8ffe15d61253d9aa6b01f5cb371be52c/MonoGame.Framework/Graphics/GraphicsDevice.cs#L131

I tried to keep these changes as safe as possible, since this code worked as-is for 99% of use cases. None of this is actually new code, it is just moved around and behind thread safety :)